proposed/middleware-meta.md

```diff
@@ -0,0 +1,99 @@
+PSR-N Meta Document
+===================
+
+1. Summary
+----------
+
+The purpose of a Middleware Interface is to make it easier to share middleware
+components between frameworks that use HTTP Message interfaces defined in PSR-7.
+
+[stackphp]: http://stackphp.com/
+
+2. Why Bother?
+--------------
+
+The general concept of a reusable request/response middleware was present in
+[StackPHP][stackphp]. Since the release of PSR-7, a number of frameworks have
+applied the StackPHP concept with a different interface signature. The signature
+for these middleware implementations has been almost universally the same.
+
+Specifying a middleware interface that uses the HTTP Message interfaces will
+enable users to publish and share middleware for any framework that is PSR-7
+compatible.
+
+3. Scope
+--------
+
+## 3.1 Goals
+
+* Provide an interface for middleware that uses HTTP Message interfaces.
+
+## 3.2 Non-Goals
+
+* This proposal does not suggest an implementation of middleware management or
+  define the sequencing mechanism for executing middleware.
+
+4. Approaches
+-------------
+
+### 4.1 Chosen Approach
+
+Based on the middleware implementations already used by frameworks that have
+adopted PSR-7, the following commonalities are observed:
+
+* The middleware is defined as a [callable][php-callable].
+* The middleware is passed 3 arguments during invocation:
+  1. A `ServerRequestInterface` implementation.
+  2. A `ResponseInterface` implemenation.
+  3. A `callable` that receives the request and response and executes the next middleware.
+* The implementation of the middleware manager is not specified as an interface.
+
+[php-callable]: http://php.net/manual/language.types.callable.php
+
+A similar interface definition can be seen in the following frameworks:
+
+* [Zend Stratigility](https://github.com/zendframework/zend-stratigility/blob/1.0.0/src/MiddlewarePipe.php#L69-L79)
+* [Relay](https://github.com/relayphp/Relay.Relay/blob/1.0.0/src/MiddlewareInterface.php#L24)
+* [Slim](https://github.com/slimphp/Slim/blob/3.4.0/Slim/MiddlewareAwareTrait.php#L66-L75)
+* [Middleman](https://github.com/mindplay-dk/middleman/blob/1.0.0/src/MiddlewareInterface.php#L24)
```

Comment:
@shadowhand 

Not sure if it's worth mentioning, but there are actually quite a bunch of projects out there using the exact same interface, while not being frameworks, they provide middleware implementing that interface.

For example, see this list right here, taken from https://github.com/refinery29/piston/pull/192:

### Interface

A bunch of projects provide and/or use exactly the same interface, for example

* [`bitexpert/adroit`](https://github.com/bitExpert/adroit)
* [`madewithlove/glue`](https://github.com/madewithlove/glue)
* [`radar/adr`](https://github.com/radarphp/Radar.Adr)
* [`relay/relay`](https://github.com/relayphp/Relay.Relay/blob/1.x/src/MiddlewareInterface.php#L42)
* [`zendframework/zend-expressive`](https://github.com/zendframework/zend-expressive/blob/2db114f9c888e5ce06c808c44c47f70324be4d5c/src/Application.php#L138)
* [`zendframework/zend-stratigility`](https://github.com/zendframework/zend-stratigility/blob/master/src/MiddlewareInterface.php#L57)
* ...

### Existing Middleware using this interface

* [`akrabat/rka-ip-address-middleware`](https://github.com/akrabat/rka-ip-address-middleware)
* [`akrabat/rka-scheme-and-host-detection-middleware`](https://github.com/akrabat/rka-scheme-and-host-detection-middleware)
* [`bear/middleware`](https://github.com/bearsunday/BEAR.Middleware)
* [`hannesvdvreken/psr7-middlewares`](https://github.com/hannesvdvreken/psr7-middlewares)
* [`los/api-problem`](https://github.com/Lansoweb/api-problem)
* [`los/los-rate-limit`](https://github.com/Lansoweb/LosRateLimit)
* [`monii/monii-action-handler-psr7-middleware`](https://github.com/monii/monii-action-handler-psr7-middleware)
* [`monii/monii-nikic-fast-route-psr7-middleware`](https://github.com/monii/monii-nikic-fast-route-psr7-middleware)
* [`monii/monii-response-assertion-psr7-middleware`](https://github.com/monii/monii-response-assertion-psr7-middleware)
* [`mtymek/blast-base-url`](https://github.com/mtymek/blast-base-url)
* [`ocramius/psr7-session`](https://github.com/Ocramius/PSR7Session)
* [`oscarotero/psr7-middlewares`](https://github.com/oscarotero/psr7-middlewares)
* [`php-middleware/block-robots`](https://github.com/php-middleware/block-robots)
* [`php-middleware/http-authentication`](https://github.com/php-middleware/http-authentication)
* [`php-middleware/log-http-messages`](https://github.com/php-middleware/log-http-messages)
* [`php-middleware/maintenance`](https://github.com/php-middleware/maintenance)
* [`php-middleware/phpdebugbar`](https://github.com/php-middleware/phpdebugbar)
* [`php-middleware/request-id`](https://github.com/php-middleware/request-id)
* [`relay/middleware`](https://github.com/relayphp/Relay.Middleware)